src/streams.rs

```diff
@@ -0,0 +1,669 @@
+        for row in &parts {
+            let mut p = StreamPendingId::default();
+            p.id = row[0].0.to_owned();
+            p.consumer = row[0].1.to_owned();
+            p.last_delivered_ms = row[0].2.to_owned();
+            p.times_delivered = row[0].3.to_owned();
+            reply.ids.push(p);
+        }
+        Ok(reply)
+    }
+}
+
+impl FromRedisValue for StreamInfoStreamReply {
+    fn from_redis_value(v: &Value) -> RedisResult<Self> {
+        let map: HashMap<String, Value> = from_redis_value(v)?;
+        let mut reply = StreamInfoStreamReply::default();
+        if let Some(v) = &map.get("last-generated-id") {
+            reply.last_generated_id = from_redis_value(v)?;
+        }
+        if let Some(v) = &map.get("radix-tree-nodes") {
+            reply.radix_tree_keys = from_redis_value(v)?;
+        }
+        if let Some(v) = &map.get("groups") {
+            reply.groups = from_redis_value(v)?;
+        }
+        if let Some(v) = &map.get("length") {
+            reply.length = from_redis_value(v)?;
+        }
+        if let Some(v) = &map.get("first-entry") {
+            reply.first_entry = StreamId::from_bulk_value(v)?;
+        }
+        if let Some(v) = &map.get("last-entry") {
+            reply.last_entry = StreamId::from_bulk_value(v)?;
+        }
+        Ok(reply)
+    }
+}
+
+impl FromRedisValue for StreamInfoConsumersReply {
+    fn from_redis_value(v: &Value) -> RedisResult<Self> {
+        let consumers: Vec<HashMap<String, Value>> = from_redis_value(v)?;
+        let mut reply = StreamInfoConsumersReply::default();
+        for map in consumers {
+            let mut c = StreamInfoConsumer::default();
+            if let Some(v) = &map.get("name") {
+                c.name = from_redis_value(v)?;
+            }
+            if let Some(v) = &map.get("pending") {
+                c.pending = from_redis_value(v)?;
+            }
+            if let Some(v) = &map.get("idle") {
+                c.idle = from_redis_value(v)?;
+            }
+            reply.consumers.push(c);
+        }
+
+        Ok(reply)
+    }
+}
+
+impl FromRedisValue for StreamInfoGroupsReply {
+    fn from_redis_value(v: &Value) -> RedisResult<Self> {
+        let groups: Vec<HashMap<String, Value>> = from_redis_value(v)?;
+        let mut reply = StreamInfoGroupsReply::default();
+        for map in groups {
+            let mut g = StreamInfoGroup::default();
+            if let Some(v) = &map.get("name") {
+                g.name = from_redis_value(v)?;
+            }
+            if let Some(v) = &map.get("pending") {
+                g.pending = from_redis_value(v)?;
+            }
+            if let Some(v) = &map.get("consumers") {
+                g.consumers = from_redis_value(v)?;
+            }
+            if let Some(v) = &map.get("last-delivered-id") {
+                g.last_delivered_id = from_redis_value(v)?;
+            }
+            reply.groups.push(g);
+        }
+        Ok(reply)
+    }
+}
```

Comment:
All these `from_redis_value` functions could use use `collect` instead of explicit for loops. Perhaps with the help of adaptors like `flat_map` in some cases

Reply:
Got it.  I have one of these replaced in the latest commit.  Shouldn't be too bad to knock out the rest.